### DISCUSSION TITLE
Can't use reexported function with same name as file in `route!`. Is this intended?

### DISCUSSION THREAD

**Discussion by arctic-alpaca** [rwf2/Rocket#1801] [Questions] [answered]
Hello,

I'm trying to use a reexported function, with the same name as the file it is contained in, in a `route!` and for some reason it doesn't compile. The same usecase seems to work with actix so I'm not sure whether there is a mistake in my code or if this is a bug in Rocket or maybe a limitation of the rust compiler.

I have a function `health_check` in `src/routes/health_check.rs` and reexport it in `src/routes/mod.rs/` with `pub use health_check::*;`. I then import it with `use crate::routes::health_check;` in `src/startup.rs` and try to use it with `routes![health_check]`.

The compiler show this error message: 
```
error[E0574]: expected struct, variant or union type, found function `health_check`
  --> src/startup.rs:22:41
   |
22 |         .mount("/health_check", routes![health_check])
   |                                         ^^^^^^^^^^^^ not a struct, variant or union type
   |
help: consider importing this struct instead
   |
1  | use crate::routes::health_check::health_check;
   |
```

If I do that, it works (after changing the visibility of `mod health_check;` in `src/routes/mod.rs/`. 

Changing the name of the function to something that's different from the `.rs` file it is contained in also makes the code compile.

This is the code: https://github.com/arctic-alpaca/zero2prod_rocket
This CI run also shows the error: https://github.com/arctic-alpaca/zero2prod_rocket/runs/3210343077?check_suite_focus=true

This is the similar actix code: https://github.com/LukeMathWalker/zero-to-production/tree/root-chapter-03-part1
In the actix code, the subscribe function body can be replaced with `HttpResponse::InternalServerError().finish()` to allow for it to compile without sqlx issues / a running database.

I'd appreciate any input why this doesn't compile.

**Comment by jebrosen** [comment, marked as answer]
No, it's not intentional. But I'm also not sure if it's too easily avoidable.

The problem does have to do with the name clash:

> I have a function `health_check` in `src/routes/health_check.rs` and reexport it in `src/routes/mod.rs/` with `pub use health_check::*;`.

At this point you would have a module, `routes`, with these items in it: `mod health_check;`, `fn health_check`, and `struct health_check`. (`struct health_check` was generated by rocket). But because that struct has the same name as the module `health_check`, `pub use health_check::*;` does not reexport it. You can see this behavior in this playground example: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=2ba98a811cf614dd722c9391143ec161 .

Incidentally, this is one of a few reasons in favor of avoiding `*` imports - `rustc` emits an error explaining the name conflict at this point, if you use `pub use health_check::health_check;`.

> I then import it with use crate::routes::health_check; in src/startup.rs and try to use it with routes![health_check].

`routes!` wanted the `struct health_check`, but it wasn't actually reexported - and then you get a less good error message.

> This is the similar actix code: https://github.com/LukeMathWalker/zero-to-production/tree/root-chapter-03-part1

In this actix-web version, it looks like you attach the routes in a way that does not use any macros. In Rocket's case, several important features (including error messages for route handlers) depend a lot on the macros and I don't think we could readily provide a similar option.

IIRC actix-web also has `#[get]` decorators as an option, and it may or may not have similar limitations.

---

For what it's worth I generally recommend not repeating names like `health_check::health_check` even when there is not a clash. In this case, I would probably have put the `fn health_check()` in `src/routes/mod.rs` without its own dedicated module because it is small, or in a `utils.rs` or `general.rs` with other routes in the same category.

**Reply by arctic-alpaca** [reply]
Thank you for your in depth answer!
